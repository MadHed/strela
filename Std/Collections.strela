module Std.Collections {
    /**
     * Contiguous storage resizable list
     */
    export class List<T> {
        var array: T[];
        var length: int;

        function init() {
            this.array = new T[](1);
            this.length = 0;
        }

        function append(value: T) {
            var newArray = new T[](this.array.length + 1);
            var i = 0;
            while(i < this.array.length) {
                newArray[i] = this.array[i];
                i = i + 1;
            }
            this.array = newArray;
            this.array[this.length] = value;
            this.length = this.length + 1;
        }

        function set(index: int, value: T) {
            this.array[index] = value;
        }

        function [](i: int): T {
            return this.array[i];
        }
    }

    /**
     * Key-Value storage
     */
    export class Map<K, V> {
        var keys: List<K>;
        var values: List<V>;

        function init() {
            this.keys = new List<K>;
            this.values = new List<V>;
        }

        function has(key: K): bool {
            return this.keyIndex(key) >= 0;
        }

        function keyIndex(key: K): i64 {
            var i: u64 = 0;
            while (i < this.keys.length) {
                if (this.keys[i] == key) return i;
                i += 1;
            }
            return -1;
        }

        function set(key: K, val: V) {
            var idx = this.keyIndex(k);
            if (idx >= 0) {
                this.values[idx] = v;
            }
            else {
                this.keys.append(k);
                this.values.append(v);
            }
        }

        function [](key: K): V {
            return this.values(this.keyIndex(kex));
        }
    }
}