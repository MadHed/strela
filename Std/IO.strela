// Copyright (c) 2018 Stephan Unverwerth
// This code is licensed under MIT license (See LICENSE for details)

module Std.IO {
    export function print(v: int): void {}
    export function print(v: float): void {}
    export function print(v: String): void {}
    export function print(v: bool): void {}

    export function println(v: int): void { print(v); print("\n"); }
    export function println(v: float): void { print(v); print("\n"); }
    export function println(v: String): void { print(v); print("\n"); }
    export function println(v: bool): void { print(v); print("\n"); }
}