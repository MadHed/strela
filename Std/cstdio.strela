// Copyright (c) 2018 Stephan Unverwerth
// This code is licensed under MIT license (See LICENSE for details)

module Std.cstdio {
    // File IO
    export external function fopen(name: String, mode: String): u32;
    export external function fclose(file: u32): i32;
    export external function feof(file: u32): i32;
    export external function fread(buffer: Ptr, size: u32, count: u32, file: u32): u32;
    export external function fwrite(buffer: Ptr, size: u32, count: u32, file: u32): u32;
    export external function fgetc(file: u32): i32;
    export external function ungetc(ch: i32, file: u32): i32;

    // Strings
    export external function strlen(str: String): i32;
    export external function itoa(val: i32, str: String, base: i32): u64;
    export external function atoi(str: String): i32;

    // Random
    export external function rand(): i32;

    // Math
    export external function sqrt(f: f64): f64;
    export external function sin(r: f64): f64;
    export external function cos(r: f64): f64;
    export external function tan(r: f64): f64;

    export external function sqrtf(f: f32): f32;
    export external function sinf(r: f32): f32;
    export external function cosf(r: f32): f32;
    export external function tanf(r: f32): f32;
}