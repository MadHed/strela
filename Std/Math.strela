// Copyright (c) 2018 Stephan Unverwerth
// This code is licensed under MIT license (See LICENSE for details)

module Std.Math {
    import Std.cstdio.*;

    export function frand(): f64 {
        return (rand() as f64 / 2147483647 as f64) - 0.5;
        //return (rand() as f64 / 65535 as f64) - 0.5;
    }

    export function clamp(f: f64): f64 {
        if (f > 1.0) return 1.0;
        if (f < 0.0) return 0.0;
        return f;
    }

    export function clamp(f: f32): f32 {
        if (f > 1.0) return 1.0;
        if (f < 0.0) return 0.0;
        return f;
    }
}