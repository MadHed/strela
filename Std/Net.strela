// Copyright (c) 2018 Stephan Unverwerth
// This code is licensed under MIT license (See LICENSE for details)

module Std.Net {
    import Std.csockets;
    import Std.cstdio.*;

    export class IPAddr {
        var a: u8;
        var b: u8;
        var c: u8;
        var d: u8;

        function init(a: u8, b: u8, c: u8, d: u8) {
            this.a = a;
            this.b = b;
            this.c = c;
            this.d = d;
        }

        function getInt(): u32 {
            var i: u32 = d + c * 256 + b * 65536 + a * 16777216;
            return i;
        }
    }

    /*
    AF_UNIX		1
    AF_INET		2
    AF_INET6	10
    */

    export class Socket {
        var handle: i32;

        function init(s: i32) {
            handle = s;
        }

        function init(ip: u32, port: u16) {
            handle = csockets.socket(2, 1, 0);
            bind(ip, port);
        }

        function init() {
            handle = csockets.socket(2, 1, 0);
        }

        function connect(ip: u32, port: u16) {
            var addr = new csockets.sockaddr_in;
            addr.family = 2;
            addr.addr = csockets.htonl(ip);
            addr.port = csockets.htons(port);
            csockets.connect(handle, addr, 16);
        }

        function readLine(): String {
            var str = "";
            var tmp = "" + 60;
            while (tmp != "\n" && csockets.recv(handle, tmp, 1, 0) > 0) {
                str = str + tmp;
            }
            return str;
        }

        function write(str: String) {
            csockets.send(handle, str, strlen(str), 16384); //MSG_NOPIPE
        }

        function bind(ip: u32, port: u16) {
            var addr = new csockets.sockaddr_in;
            addr.family = 2;
            addr.addr = csockets.htonl(ip);
            addr.port = csockets.htons(port);

            csockets.bind(handle, addr, 16);
        }

        function listen() {
            csockets.listen(handle, 1);
        }

        function accept(): Socket {
            var addr = new csockets.sockaddr_in;
            var len: u32 = 0;
            var newsock = csockets.accept(handle, addr, len);
            return new Socket(newsock);
        }

        function shutdown() {
            csockets.shutdown(handle, 0);
        }

        function close() {
            csockets.close(handle);
        }
    }
}