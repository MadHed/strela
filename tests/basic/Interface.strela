module Interface {
    import Std.IO.*;

    interface ForceUser {
        function getName(): String;
        function foo(i: int): String;
    }

    class Vader {
        var name: String;
        
        function init() {
            this.name = "Darth Vader";
        }

        function getName(): String {
            return this.name;
        }

        function foo(i: int): String {
            return "I am your father.";
        }
    }

    class Yoda {
        var name: String;

        function init() {
            this.name = "Master Yoda";
        }

        function getName(): String {
            return this.name;
        }

        function foo(i: int): String {
            return "Do or do not. There is no try.";
        }
    }

    function printFoo(user: ForceUser) {
        print("My name is ");
        print(user.getName());
        println(".");
        println(user.foo(1));
    }

    function main(args: String[]): int {
        var vader = new Vader;
        var yoda = new Yoda;

        printFoo(vader);
        printFoo(yoda);

        return 0;
    }
}