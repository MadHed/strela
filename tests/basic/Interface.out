My name is Darth Vader.
I am your father.
My name is Master Yoda.
Do or do not. There is no try.
